dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java

```diff
@@ -0,0 +1,110 @@
+
+    @Autowired
+    private WorkFlowLineageMapper workFlowLineageMapper;
+
+    public Map<String, Object> queryWorkFlowLineageByName(String workFlowName, int projectId) {
+        Map<String, Object> result = new HashMap<>(5);
+        List<WorkFlowLineage> workFlowLineageList = workFlowLineageMapper.queryByName(workFlowName, projectId);
+        result.put(Constants.DATA_LIST, workFlowLineageList);
+        putMsg(result, Status.SUCCESS);
+        return result;
+    }
+
+    private List<WorkFlowRelation> getWorkFlowRelationRecursion(Set<Integer> ids, List<WorkFlowRelation> workFlowRelations,Set<Integer> sourceIds) {
+        for (int id : ids) {
+            sourceIds.addAll(ids);
+            List<WorkFlowRelation> workFlowRelationsTmp = workFlowLineageMapper.querySourceTarget(id);
+            if (workFlowRelationsTmp != null && !workFlowRelationsTmp.isEmpty()) {
+                Set<Integer> idsTmp = new HashSet<>();
+                for (WorkFlowRelation workFlowRelation:workFlowRelationsTmp) {
+                    if (!sourceIds.contains(workFlowRelation.getTargetWorkFlowId())) {
+                        idsTmp.add(workFlowRelation.getTargetWorkFlowId());
+                    }
+                }
+                workFlowRelations.addAll(workFlowRelationsTmp);
+                getWorkFlowRelationRecursion(idsTmp, workFlowRelations,sourceIds);
+            }
+        }
+        return workFlowRelations;
+    }
+
+    public Map<String, Object> queryWorkFlowLineageByIds(Set<Integer> ids,int projectId) {
+        Map<String, Object> result = new HashMap<>(5);
+        List<WorkFlowLineage> workFlowLineageList = workFlowLineageMapper.queryByIds(ids, projectId);
+        Map<String, Object> workFlowLists = new HashMap<>(5);
+        Set<Integer> idsV = new HashSet<>();
+        if (ids == null || ids.isEmpty()) {
+            for (WorkFlowLineage workFlowLineage:workFlowLineageList) {
+                idsV.add(workFlowLineage.getWorkFlowId());
+            }
+        } else {
+            idsV = ids;
+        }
+        List<WorkFlowRelation> workFlowRelations = new ArrayList<>();
+        Set<Integer> sourceIds = new HashSet<>();
+        getWorkFlowRelationRecursion(idsV, workFlowRelations, sourceIds);
+
+        Set<Integer> idSet = new HashSet<>();
+        //If the incoming parameter is not empty, you need to add downstream workflow detail attributes
+        if (ids != null && !ids.isEmpty()) {
+            for (WorkFlowRelation workFlowRelation : workFlowRelations) {
+                idSet.add(workFlowRelation.getTargetWorkFlowId());
+            }
+            for (int id : ids) {
+                idSet.remove(id);
+            }
+            if (!idSet.isEmpty()) {
+                workFlowLineageList.addAll(workFlowLineageMapper.queryByIds(idSet, projectId));
+            }
+        }
+
+        workFlowLists.put("workFlowList",workFlowLineageList);
```

Comment:
It's best not to show magic numbers

Reply:
Replaced with `Constants.WORKFLOW_LIST` and `Constants.WORKFLOW_RELATION_LIST`